src/utilities/modules/vaccines.js

```diff
@@ -0,0 +1,170 @@
+    if (!sensorLogsByLocation[location.id]) sensorLogsByLocation[location.id] = [sensorLog];
+    else sensorLogsByLocation[location.id].push(sensorLog);
+  });
+
+  Object.values(sensorLogsByLocation).forEach(logsForLocation => {
+    logsForLocation.forEach(sensorLog => {
+      const { isInBreach } = sensorLog;
+      const { length: stackLength } = sensorLogStack;
+      // If the stack is empty, just push this sensorLog.
+      if (stackLength === 0) sensorLogStack.push(sensorLog);
+      // Any breached sensorLogs are always pushed onto the stack.
+      else if (isInBreach) sensorLogStack.push(sensorLog);
+      // If both head of the stack and this sensorLog are unbreached,
+      // replace the head. This can occurs when the length is 1. Two
+      // sequential unbreached logs are never pushed.
+      else if (!sensorLogStack[stackLength - 1].isInBreach && !isInBreach) {
+        sensorLogStack.pop();
+        sensorLogStack.push(sensorLog);
+        // In all other cases, a breach has been found. Push the last sensorlog,
+        // store the breach and reset the stack. Push this sensorLog back
+        // onto the stack as the new potential delimiter for the next breach.
+      } else {
+        sensorLogStack.push(sensorLog);
+        breaches.push([sensorLogStack]);
+        sensorLogStack = [];
+        sensorLogStack.push(sensorLog);
+      }
+    });
+  });
+
+  // Push any remaining sensorlogs. If any are left, they form a breach
+  // with no delimiting non-breached sensorlog.
+  if (sensorLogStack.length > 0) breaches.push(sensorLogStack);
+
+  // Create Realm.result objects for each breach.
+  return breaches.map(breach => {
+    const sensorIds = breach.map(log => log.id);
+    const queryString = sensorIds.map((_, i) => `id = $${i}`).join(' OR ');
+    return database.objects('SensorLog').filtered(queryString, ...sensorIds);
+  });
+};
+
+/**
+ * Extracts all ItemBatches which are related to the passed sensorLogs.
+ * Groups all ItemBatch objects by Item and calculates the duration of
+ * time each ItemBatch was present during these logs.
+ *
+ * Option parameters itemBatch and item are used to filter the ItemBatches
+ * for these sensorLogs to only search for ItemBatch records which are
+ * related to the Item xor the ItemBatch. Item takes priority over
+ * ItemBatch if both are passed.
+ *
+ * Method iterates through each batch, grouping batches by Item. Will store
+ * values on the first encounter of a batch, otherwise will just increment
+ * the duration.
+ *
+ * @param  {Realm.results/array} sensorLogs
+ * @param  {ItemBatch}           itemBatch
+ * @param  {Item}                item
+ * @return {Array<Object>} array of objects, see example below.
+ * [
+ *  {
+ *     item,
+ *     batches: [ { id, code, expiryDate, enteredDate, duration, totalQuantity }]
+ *  }
+ *  ...
+ * ]
+ */
+export const sensorLogsExtractBatches = ({ sensorLogs = [], itemBatch, item } = {}) => {
+  const groupedBatches = {};
```

Comment:
once again change not required, you can do this however:
filter sensorLogs by item or batch
for each sensorLogs, if groupedBatches,item.batch exists return 
otherwise add it and fine 'duration' by :
if not find duration, by sensorLogs.filtered by batch .max('timestamp') - .min('timestamp')

Reply:
1+ (3+2+1-5+2-3 ) - b8919e997007a723bf88c9af19ab8e12a5807416